### ISSUE TITLE
Questions regarding workflow for novel modifications 

### ISSUE THREAD

**Issue by drivenbyentropy** [nanoporetech/remora#62]
Hello,

We are currently attempting to train a remora model for a modification that is currently not present in the list of pretrained models. To this end, we have generated reads containing canonical bases only, as well as a treated sample with contains reads with modified nucleotides. 

My current workflow is as follows:
1) Convert fast5 to pod using ```pod5-convert-from-fast5```
2) Use the latest guppy version to call bases with```-move_out```, ```--bam_out```, ```--align_ref``` flags for treatment and control respectively
3) Merge files into a single bam file for treatment and control respectively
4) Prepare control and treatment datasets with ```remora dataset prepare```
```
remora dataset prepare \
       "fast5_control_pod5/output.pod5" \
       "fast5_control_merged.bam" \
       --output-remora-training-file "fast5_control_chunks.npz" \
       --mod-base-control 

remora dataset prepare \
       "fast5_treatment_pod5/output.pod5" \
       "fast5_treatment_merged.bam" \
       --output-remora-training-file "fast5_treatment_chunks.npz" \
       --mod-base m NewMod \
```
5) Merge chunks
```
remora dataset merge \
       --input-dataset "fast5_control_chunks.npz" 14_000_000 \
       --input-dataset "fast5_treatment_chunks.npz" 14_000_000 \
       --output-dataset "fast5_combined_chunks.npz"
```
6) Train the model
```
remora model train \
       "fast5_combined_chunks.npz" \
       --model "[PATH_TO_REMORA]/models/ConvLSTM_w_ref.py" \
       --device 0 \
       --output-path "train_results"
```

While from a computational perspective, this workflow appears to work, I do have some conceptual questions that I was hoping you could shed some light on:
1) I have read in several posts (e.g. [here](github.com/nanoporetech/remora/issues/37)) that in order to successfully train a model, the modified sample should be fully modified. Is this correct? How would training be affected if some training data contains pairs in which both loci are not modified? Do you have any recommendations on how to preprocess reads for which only a portion of the genomic loci are modified?

2) Regarding the target base, it is not clear to me, at which point this can be specified. I initially assumed this to correspond to the ```mod-base``` parameter, but documentation is scarce. The help suggests this parameter to be for ```Modified base information. Example: `--mod-base m 5mC` (default: None)```, but from my searches I concluded that these are simply to specify placeholder IDs for BAM and model output files produced downstream. Specifically, assuming that the modification primarily targets ```Ts```, how can this information be passed to ```remora```? In addition, if the affinity of the modification is towards two bases, do you recommend to train two separate models in this case (see [this issue](github.com/nanoporetech/remora/issues/15))? 

3) Instead of specifying a fixed maximum number of chunks during the merge phase, does it make sense to include all chunks (possibly resulting in an unbalanced dataset) and use the ```--balanced-batch``` flag at the time of training?

Thank you very much in advance for your answers!

**Comment by drivenbyentropy** [comment]
**Edit regarding question 2**: After digging into the source code, I believe my understanding of the ```--mod-base``` option was not accurate. The first parameter seems to be the single letter code for the target modified bases, whereas the second one is a long name space holder. I was confused by the lower case ```m``` but now believe this corresponds to "CT" as defined in ```util.py```  Is this correct? 

```
SINGLE_LETTER_CODE = {
    "A": "A",
    "C": "C",
    "G": "G",
    "T": "T",
    "B": "CGT",
    "D": "AGT",
    "H": "ACT",
    "K": "GT",
    "M": "AC",
    "N": "ACGT",
    "R": "AG",
    "S": "CG",
    "V": "ACG",
    "W": "AT",
    "Y": "CT",
}
```

**Comment by marcus1487** [comment]
There are a number of points to cover here. I'll try to cover them one at a time, but please comment on anything still outstanding. 

The `--mod-base` argument takes a single letter and "long" name for the modified base. The `m` code is the community standard for 5mC (see [SAM tag specifications](https://samtools.github.io/hts-specs/SAMtags.pdf), "Base Modifications" Section). This single letter is required in order to specify the modified base in the output SAM/BAM format. We will add a link to this document in the next release.

> I have read in several posts (e.g. [here](https://github.com/nanoporetech/remora/issues/github.com/nanoporetech/remora/issues/37)) that in order to successfully train a model, the modified sample should be fully modified. Is this correct? How would training be affected if some training data contains pairs in which both loci are not modified? Do you have any recommendations on how to preprocess reads for which only a portion of the genomic loci are modified?

Yes, the modified sample should be completely modified at the specified locations (either via the `--motif` or `--focus-reference-positions` arguments). Modified bases are quite tricky so in practice 100% modified bases are not often possible. The arguments provided are a start toward assisting training in scenarios when labels may be incorrect. We are working on more advanced methods to handle this situation and will make the best practices the default in remora as we continue our research.

> In addition, if the affinity of the modification is towards two bases, do you recommend to train two separate models in this case (see [this issue](https://github.com/nanoporetech/remora/issues/github.com/nanoporetech/remora/issues/15))?

Affinity towards multiple bases should be handled by the Remora and can be used for inference via Bonito, though this is not tested so there may be edge cases which were not considered. The ``--motif`` argument will specify locations at which a model can be applied. Specifying an ambiguous base as the focus position of the motif should work for this. Again this is untested so there may be some issues with model against multiple canonical bases and you may have to train multiple models for now.

> Instead of specifying a fixed maximum number of chunks during the merge phase, does it make sense to include all chunks (possibly resulting in an unbalanced dataset) and use the --balanced-batch flag at the time of training?

These are both options. Also the merge function has a balance function to balance the whole dataset. We have not definitively determined a best practice across many samples for this setting at this time. In the next release we will also make ``all`` a specific option for the merge function to avoid choosing some arbitrarily large number.

**Comment by marcus1487** [comment]
This is not provided in the current interface. We could consider adding such an interface in the future, but this would be considered a research task at the moment. I would recommend looking into using the API directly for such applications. This [point in the code could be a good starting point](https://github.com/nanoporetech/remora/blob/master/src/remora/prepare_train_data.py#L88). This is where the "focus positions" are annotated in the code. You would need to add the table of read id/positions here (either reference positions or basecall positions) and add this logic. The next release should make this a bit easier as well.

**Comment by marcus1487** [comment]
I'm not seeing how this could happen directly from a `remora dataset prepare` command. Can you pos the full dataset prepare command and the results of `remora dataset inspect` on the resulting dataset (before merging with other datasets or any other manipulation).

**Comment by solvedgamecompany** [comment]
Here are the comands and their outputs:

```
remora dataset prepare merged_6ma.pod5 basecalls_6ma.bam  --output-remora-training-file m6a_chunks_AG_motif.npz --mod-base a 6mA --motif AG 0  --num-extract-alignment-workers 8 --num-extract-chunks-workers 8 --num-reads 2000000
```

```
 remora dataset inspect m6a_chunks_AG_motif.npz
```
Output:
```
Dataset summary:
               num chunks : 28422203
       label distribution : Counter({1: 28422203})
                base_pred : False
                mod_bases : a
           mod_long_names : ('6mA',)
       kmer_context_bases : (4, 4)
            chunk_context : (50, 50)
                   motifs : [('AG', 0)]
 chunk_extract_base_start : False
     chunk_extract_offset : 0
          sig_map_refiner : No Remora signal refine/map settings loaded
```

**Comment by mvolar** [comment]
Here are the comands and their outputs:

```
remora dataset prepare merged_6ma.pod5 basecalls_6ma.bam  --output-remora-training-file m6a_chunks_AG_motif.npz --mod-base a 6mA --motif AG 0  --num-extract-alignment-workers 8 --num-extract-chunks-workers 8 --num-reads 2000000
```

```
 remora dataset inspect m6a_chunks_AG_motif.npz
```
Output:
```
Dataset summary:
               num chunks : 28422203
       label distribution : Counter({1: 28422203})
                base_pred : False
                mod_bases : a
           mod_long_names : ('6mA',)
       kmer_context_bases : (4, 4)
            chunk_context : (50, 50)
                   motifs : [('AG', 0)]
 chunk_extract_base_start : False
     chunk_extract_offset : 0
          sig_map_refiner : No Remora signal refine/map settings loaded
```
